### PR TITLE
fs/procfs: Supports any number of thread displays

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
After the number of threads exceeds the array size, it will not be displayed.
Any number of threads can be displayed using dynamic adaptation

### CODE REVIEW

File: fs/procfs/fs_procfs.c [hunk 1]
```diff
@@ -627,12 +636,18 @@ static int procfs_opendir(FAR struct inode *mountpt, FAR const char *relpath,
...
        * dirent structure.
        */
 
+#ifndef CONFIG_FS_PROCFS_EXCLUDE_PROCESS
+      nxsched_foreach(procfs_thread_number, &num);
+#endif
+
       level0 = (FAR struct procfs_level0_s *)
-         kmm_zalloc(sizeof(struct procfs_level0_s));
+         kmm_zalloc(sizeof(struct procfs_level0_s) + sizeof(pid_t) * num) ;
```
Comment (anchao) [inline fs/procfs/fs_procfs.c:650]:
```suggestion
         kmm_zalloc(sizeof(struct procfs_level0_s) + sizeof(pid_t) * g_npidhash) ;
```

**Review by anchao — CHANGES_REQUESTED** [review]

File: fs/procfs/fs_procfs.c (hunk 1 above)
Comment (xiaoxiang781216) [reply fs/procfs/fs_procfs.c:650]: g_npidhash may large than the real number of thread

**Review by xiaoxiang781216 — APPROVED** [review]

File: fs/procfs/fs_procfs.c (hunk 1 above)
Comment (anchao) [reply fs/procfs/fs_procfs.c:650]: Yes, but It is fine iflarger than the real task. line 650 will foreach task again.

File: fs/procfs/fs_procfs.c [hunk 2]
```diff
@@ -309,8 +309,8 @@ struct procfs_level0_s
   /* Our private data */
 
   uint8_t lastlen;                       /* length of last reported static dir */
-  pid_t pid[CONFIG_FS_PROCFS_MAX_TASKS]; /* Snapshot of all active task IDs */
   FAR const char *lastread;              /* Pointer to last static dir read */
+  pid_t pid[0];                          /* Snapshot of all active task IDs */
```
Comment (pkarashchenko) [inline fs/procfs/fs_procfs.c:313]: can we make it `pid_t pid[1];` as `pid_t pid[0];` is C89 incompatible?

File: fs/procfs/fs_procfs.c (hunk 1 above)
Comment (xiaoxiang781216) [reply fs/procfs/fs_procfs.c:650]: Ok

Comment (Gary-Hobson) [reply fs/procfs/fs_procfs.c:649]: Done

File: fs/procfs/fs_procfs.c [hunk 3]
```diff
@@ -633,12 +634,20 @@ static int procfs_opendir(FAR struct inode *mountpt, FAR const char *relpath,
...
   if (!relpath || relpath[0] == '\0')
     {
+      size_t num = 0;
+
       /* The path refers to the top level directory.  Allocate the level0
        * dirent structure.
        */
 
+#ifndef CONFIG_FS_PROCFS_EXCLUDE_PROCESS
+      irqstate_t flags = enter_critical_section();
```
Comment (xiaoxiang781216) [inline fs/procfs/fs_procfs.c:644]: don't need hold the critical section

Comment (Gary-Hobson) [reply fs/procfs/fs_procfs.c]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**